container credentials

till opened the pull request:
- Chore: add a snapshot target
- Update: support credentials for private containers

Resolves: #788

till commented on pkg/runner/run_context.go:
```diff
@@ -102,6 +102,14 @@ func (rc *RunContext) startJobContainer() common.Executor {
+	username := rc.Config.Secrets["DOCKER_USERNAME"]
+	password := rc.Config.Secrets["DOCKER_PASSWORD"]
+
+	if rc.Run.Job().Container().Credentials["username"] != "" {
```
Just reading this again. I bet I should not assume `username` is in the map? I'll fix later.

catthehacker replied:
```suggestion
	if len(rc.Run.Job().Container().Credentials) == 2 {
```
this should work

catthehacker replied:
But it would be good to check if credentials allow interpolated values like:
`username: ${{ env.USER }}`
(I assume they do)

till replied:
@catthehacker Yes, good call, this is common:

```
username: token
password: ${{ github.token }}
```

Or think `${{ secrets.foo }}`.

till commented:
Fixed the check as suggested ✅ . I think I added the expressions as well. Not entire sure how to write a test for the `run_context` 👹. Anyone have any hints?

catthehacker commented:
context should be inherited from function which is calling it

till commented on pkg/runner/run_context.go:
```diff
@@ -779,3 +781,38 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
+	username = container.Credentials["username"]
+	password = container.Credentials["password"]
+
+	rawLogger := common.Logger(context.TODO()).WithField("raw_output", true)
```
@catthehacker Are you referring to this? I had no idea what to stick there? `context.Background()`?

catthehacker replied:
Example: https://github.com/nektos/act/blob/96cf907c5e1f4761ef6d91bb8e22487747a77f6b/pkg/runner/step_context.go#L161

In your case it would be:

```go
username, password := rc.handleCredentials(ctx)
[...]
func (rc *RunContext) handleCredentials(ctx context.Context) (string, string) {
[...]
```
```suggestion
	rawLogger := common.Logger(ctx).WithField("raw_output", true)
```

till replied:
I've been trying this out locally. It's not as easy. There is no `context.Context` before the closure in `startJobContainer()`. Also tried moving my code into the closure:

```
// changed the signature
func handleCredentials(rc *RunContext, ctx context.Context)
```

Ultimately panics with:

```

panic: test timed out after 30s

goroutine 500 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1788 +0x8e
created by time.goFunc
	/usr/local/Cellar/go/1.17.1/libexec/src/time/sleep.go:180 +0x31

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0001a4b60, {0x181e3ff, 0xbe2407ede71}, 0x184d310)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1307 +0x375
testing.runTests.func1(0xc0003d2ff0)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1598 +0x6e
testing.tRunner(0xc0001a4b60, 0xc0002dfce0)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1259 +0x102
testing.runTests(0xc00039c500, {0x1e0bc40, 0x12, 0x12}, {0x10894ed, 0x181fe77, 0x1e151e0})
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1596 +0x43f
testing.(*M).Run(0xc00039c500)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1504 +0x51d
main.main()
	_testmain.go:135 +0x1f5

goroutine 135 [chan receive]:
testing.(*T).Run(0xc0004c2340, {0x181b6e6, 0x1}, 0xc0002244d0)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1307 +0x375
github.com/nektos/act/pkg/runner.runTestJobFile({0x1930818, 0xc000124008}, 0x0, {{0x181c00f, 0x8}, {0x181b6e6, 0x7}, {0x181a11f, 0x4}, {0x1911410, ...}, ...})
	/Users/till/Documents/workspaces/luzilla/act/pkg/runner/runner_test.go:55 +0xdf
github.com/nektos/act/pkg/runner.TestRunEvent(0xc0004c2340)
	/Users/till/Documents/workspaces/luzilla/act/pkg/runner/runner_test.go:134 +0x387
testing.tRunner(0xc0004c2340, 0x184d310)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1306 +0x35a

goroutine 393 [chan receive]:
github.com/nektos/act/pkg/common.NewParallelExecutor.func1({0x1930818, 0xc000124008})
	/Users/till/Documents/workspaces/luzilla/act/pkg/common/executor.go:111 +0x15a
github.com/nektos/act/pkg/common.Executor.Then.func1({0x1930818, 0xc000124008})
	/Users/till/Documents/workspaces/luzilla/act/pkg/common/executor.go:146 +0x16e
github.com/nektos/act/pkg/runner.runTestJobFile.func1(0x0)
	/Users/till/Documents/workspaces/luzilla/act/pkg/runner/runner_test.go:77 +0x393
testing.tRunner(0xc000001d40, 0xc0002244d0)
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/Cellar/go/1.17.1/libexec/src/testing/testing.go:1306 +0x35a

goroutine 518 [select]:
net/http.(*persistConn).roundTrip(0xc00040b0e0, 0xc00039a480)
	/usr/local/Cellar/go/1.17.1/libexec/src/net/http/transport.go:2614 +0x97d
net/http.(*Transport).roundTrip(0xc000235cc0, 0xc0001c4b00)
	/usr/local/Cellar/go/1.17.1/libexec/src/net/http/transport.go:594 +0x7d1
net/http.(*Transport).RoundTrip(0x30, 0x191d900)
	/usr/local/Cellar/go/1.17.1/libexec/src/net/http/roundtrip.go:18 +0x19
net/http.send(0xc0001c4b00, {0x191d900, 0xc000235cc0}, {0x1801d60, 0x1, 0x0})
	/usr/local/Cellar/go/1.17.1/libexec/src/net/http/client.go:252 +0x5d8
net/http.(*Client).send(0xc0006131d0, 0xc0001c4b00, {0x59, 0x14, 0x0})
	/usr/local/Cellar/go/1.17.1/libexec/src/net/http/client.go:176 +0x9b
net/http.(*Client).do(0xc0006131d0, 0xc0001c4b00)
```

I just rebased... maybe `master` is broken?

panekj replied:
this works for me

```diff
diff --git a/pkg/runner/run_context.go b/pkg/runner/run_context.go
index c37cbb9..eeff457 100755
--- a/pkg/runner/run_context.go
+++ b/pkg/runner/run_context.go
@@ -102,9 +102,8 @@ func (rc *RunContext) startJobContainer() common.Executor {
 	image := rc.platformImage()
 	hostname := rc.hostname()
 
-	username, password := rc.handleCredentials()
-
 	return func(ctx context.Context) error {
+		username, password := rc.handleCredentials(ctx)
 		rawLogger := common.Logger(ctx).WithField("raw_output", true)
 		logWriter := common.NewLineWriter(rc.commandHandler(ctx), func(s string) bool {
 			if rc.Config.LogOutput {
@@ -782,7 +781,7 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
 	return "", false
 }
 
-func (rc *RunContext) handleCredentials() (string, string) {
+func (rc *RunContext) handleCredentials(ctx context.Context) (string, string) {
 	username := rc.Config.Secrets["DOCKER_USERNAME"]
 	password := rc.Config.Secrets["DOCKER_PASSWORD"]
 
@@ -802,7 +801,7 @@ func (rc *RunContext) handleCredentials() (string, string) {
 	username = container.Credentials["username"]
 	password = container.Credentials["password"]
 
-	rawLogger := common.Logger(context.TODO()).WithField("raw_output", true)
+	rawLogger := common.Logger(ctx).WithField("raw_output", true)
 
 	ee := rc.NewExpressionEvaluator()
 	username, _, err := ee.Evaluate(username)
```

catthehacker replied:
this works for me

```diff
diff --git a/pkg/runner/run_context.go b/pkg/runner/run_context.go
index c37cbb9..eeff457 100755
--- a/pkg/runner/run_context.go
+++ b/pkg/runner/run_context.go
@@ -102,9 +102,8 @@ func (rc *RunContext) startJobContainer() common.Executor {
 	image := rc.platformImage()
 	hostname := rc.hostname()
 
-	username, password := rc.handleCredentials()
-
 	return func(ctx context.Context) error {
+		username, password := rc.handleCredentials(ctx)
 		rawLogger := common.Logger(ctx).WithField("raw_output", true)
 		logWriter := common.NewLineWriter(rc.commandHandler(ctx), func(s string) bool {
 			if rc.Config.LogOutput {
@@ -782,7 +781,7 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
 	return "", false
 }
 
-func (rc *RunContext) handleCredentials() (string, string) {
+func (rc *RunContext) handleCredentials(ctx context.Context) (string, string) {
 	username := rc.Config.Secrets["DOCKER_USERNAME"]
 	password := rc.Config.Secrets["DOCKER_PASSWORD"]
 
@@ -802,7 +801,7 @@ func (rc *RunContext) handleCredentials() (string, string) {
 	username = container.Credentials["username"]
 	password = container.Credentials["password"]
 
-	rawLogger := common.Logger(context.TODO()).WithField("raw_output", true)
+	rawLogger := common.Logger(ctx).WithField("raw_output", true)
 
 	ee := rc.NewExpressionEvaluator()
 	username, _, err := ee.Evaluate(username)
```

till replied:
OK, I have the same. I'll commit in a couple minutes and let the test run online.

KnisterPeter reviewed (commented):
@till  Code looks fine :tada: 
Is there a way to add an integration test for this? Might be hard, but currently there is only a test which validates that the workflow could be parsed correctly.

catthehacker commented:
I've already rewritten the function, will update this PR in about 30 minutes

catthehacker commented:
https://github.com/Luzilla/act/pull/1

ZauberNerd commented on pkg/runner/run_context.go:
```diff
@@ -779,3 +784,37 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
+
+func (rc *RunContext) handleCredentials() (username, password string, err error) {
+	// TODO: remove below 2 lines when we can release act with breaking changes
+	username = rc.Config.Secrets["DOCKER_USERNAME"]
```
These secrets are used elsewhere too:
When the runner image that act is using `act -P ubuntu-latest=my.registry/some/image:latest` lives on a private registry that requires auth, act will pull it using the authentication from the secrets.
And since this is `act` internals, I don't think `container.credentials` will be sufficient.
https://github.com/nektos/act/blob/6ebcac37710ae5d4eb87f63f31d7e1be8fe3d217/pkg/container/docker_pull.go#L80-L97
Although I have also created a PR (https://github.com/nektos/act/pull/891) which uses the docker/cli package to read the registry authentication from the local docker config (if it exists).

catthehacker replied:
I'm afraid we can't change current behaviour for now

till commented:
> https://github.com/Luzilla/act/pull/1

I merged but will take a closer look tomorrow or Friday.

catthehacker commented on pkg/runner/run_context.go:
```diff
@@ -779,3 +784,37 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
+		return
+	}
+
+	if len(container.Credentials) != 2 {
```
```suggestion
	if container.Credentials != nil && len(container.Credentials) != 2 {
```
my bad, forgot to check stuff

catthehacker commented on pkg/runner/run_context.go:
```diff
@@ -779,3 +784,37 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
+	password = rc.Config.Secrets["DOCKER_PASSWORD"]
+
+	container := rc.Run.Job().Container()
+	if container == nil {
```
```suggestion
	if container == nil && container.Credentials == nil {
```

till commented on pkg/runner/run_context.go:
```diff
@@ -836,3 +841,37 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
+
+	if container.Credentials != nil && len(container.Credentials) != 2 {
+		err = fmt.Errorf("invalid property count for key 'credentials:'")
+		return
```
This is wrong?

till commented on pkg/runner/run_context.go:
```diff
@@ -836,3 +841,37 @@ func (rc *RunContext) localCheckoutPath() (string, bool) {
+	var ok bool
+	if username, ok = ee.InterpolateWithStringCheck(container.Credentials["username"]); !ok {
+		err = fmt.Errorf("failed to interpolate container.credentials.username")
+		return
```
I may be missing something, but do any of these returns work? 😬

till replied:
Learned about _named returns_ from a colleague today. :D Thanks very much, @catthehacker!

catthehacker replied:
Sorry, just noticed the comments
